### PR TITLE
Autosave changes to paperTitle when navigating to new page/view

### PR DESCRIPTION
**Pull request by neo-tahi**
References: https://www.pivotaltracker.com/story/show/85970300

## Pull Request Checklist
- [x] ask yourself: is this PR complicated enough that you should write
      integration tests?
- [x] ask yourself: do you feel that things are well abstracted?
- [x] do a self code review (read through `git diff` or the github UI)
- [x] complete all relevant todos in your code
- [x] self QA (make sure the app still works, along with your change)
- [ ] get review from a team-member with knowledge of this codebase area
- [ ] **team-member said ship it! (:ship: it)**
- [ ] self QA on staging
- [ ] do any changes need to be made to prod before a deploy? List them in
      the `deployer checklist`
- [ ] the tests are passing for this branch on semaphore
- [ ] merge! Celebrate your success!

## Reviewer Checklist
- [ ] is syntax consistent with rest of code-base?
- [ ] are database access patterns relatively efficient?
- [ ] are client-side access patterns
- [ ] ember style

## Deployer Checklist
Changes to Tahi are deployed automatically to staging whenever tests for
the `master` branch are passing. These items really only apply for deploys
to `currents`, and `demo` (we don't have a production yet).
- [ ] **>>list changes here that need to be made to prod before a deploy<<**
- [ ] ideally these changes can be made by running a rake task or script
- [ ] you are deploying in the morning
- [ ] you are deploying on Monday, Tuesday, Wednesday, or Thursday.
- [ ] you are deploying your own code
- [ ] did the deploy
- [ ] you are watching the BugSnag slack channel and/or `heroku logs --tail`
- [ ] you announced the deploy with a few words summary to chatrooms
- [ ] you have opened https://dashboard.heroku.com/apps for rollback

### CODE REVIEW

**Comment by neo-tahi** [comment]
@tennety 

review plx?

— AJP DT

File: app/assets/javascripts/initializers/error_handler.js.coffee [hunk 1]
```diff
@@ -15,9 +15,13 @@ ETahi.initializer
...
     Ember.onerror = (error) ->
       logError("\n" + error.message + "\n" + error.stack + "\n")
-      window.ErrorNotifier.notify(error, "Uncaught Ember Error")
+      # TODO FIX THIS
+      # window.ErrorNotifier.notify(error, "Uncaught Ember Error")
       if ETahi.environment == 'development'
         throw error
+      # TODO FIX THIS. Env is not set in QUnit
+      else if ETahi.environment == undefined
+        throw error
```
Comment (neo-tahi) [inline app/assets/javascripts/initializers/error_handler.js.coffee]: gotta sort out error_handler

File: app/views/shared/_bugsnag_javascript.html.haml [hunk 2]
```diff
@@ -1,3 +1,4 @@
+- # TODO REMOVE JS FROM HERE AND MOVE IT INTO ERROR_HANDLER.JS.COFFE
 - if ENV['BUGSNAG_JAVASCRIPT_API_KEY']
   %script{ src:"//d2wy8f7a9ursnm.cloudfront.net/bugsnag-2.min.js",
     "data-apikey" => "#{ENV["BUGSNAG_JAVASCRIPT_API_KEY"]}" }
```
Comment (neo-tahi) [inline app/views/shared/_bugsnag_javascript.html.haml]: it is okay to have this script tag here, but not okay to have the rest of the JS. It should go in error_handler.js

File: app/assets/javascripts/controllers/paper_edit_controller.js.coffee [hunk 3]
```diff
@@ -64,19 +64,21 @@ ETahi.PaperEditController = ETahi.BasePaperController.extend
...
-      else #locking
+      else #locking -> Editing Paper (locking others out)
         @set('lockedBy', @getCurrentUser())
         @get('model').save().then (paper) =>
           @send('startEditing')
           @set('saveState', false)
 
     savePaper: ->
+      # TODO return if ETahi.environment == 'test'
+      return if ETahi.environment == undefined
```
Comment (neo-tahi) [inline app/assets/javascripts/controllers/paper_edit_controller.js.coffee]: No good option other than this: https://github.com/Tahi-project/tahi/commit/9e1f9973c5faea6f28ac5ac85d7f097007a9b212

Fix the environment check on this line.

File: app/assets/javascripts/controllers/paper_edit_controller.js.coffee [hunk 4]
```diff
@@ -64,19 +64,20 @@ ETahi.PaperEditController = ETahi.BasePaperController.extend
...
           @set('saveState', true)
-      else #locking
+      else #locking -> Editing Paper (locking others out)
         @set('lockedBy', @getCurrentUser())
         @get('model').save().then (paper) =>
           @send('startEditing')
           @set('saveState', false)
 
     savePaper: ->
+      return if ETahi.environment == 'test'
```
Comment (rizwanreza) [inline app/assets/javascripts/controllers/paper_edit_controller.js.coffee]: I'm usually wary of these checks on environment in code, why did we have to do this?

**Comment by rizwanreza** [comment]
Looks good, other than the one comment I have.

File: app/assets/javascripts/controllers/paper_edit_controller.js.coffee (hunk 4 above)
Comment (annapetry) [inline app/assets/javascripts/controllers/paper_edit_controller.js.coffee]: We call ```saveTitleChanges```, which eventually calls ```savePaper``` on navigating away from the paper edit screen.  This works great in the app, but in testing, ```ETahi.reset()``` is called between individual test setup & teardown.  This results in an 'inFlight' model which blows up when attempting to save.  We were able to bypass this blow up in testing by checking for the  ```ETahi.environment``` variable and returning before the save was attempted.

Thoughts are appreciated.

File: app/assets/javascripts/initializers/error_handler.js.coffee [hunk 5]
```diff
@@ -9,17 +9,26 @@ ETahi.initializer
       e = new Error(msg)
       console.log(e.stack || e.message)
 
+    # TODO investigate this more
```
Comment (afomi) [inline app/assets/javascripts/initializers/error_handler.js.coffee]: investigate what more?

**Pull request merged**